### DISCUSSION TITLE
Stateless Tofu

### DISCUSSION THREAD

**Discussion by abitrolly** [opentofu/opentofu#1457] [Ideas]
I want to update YouTube video descriptions collaboratively using YAML file from a GitLab repo. It is possible to do this with custom CI/CD script, but I thought it could be a fun project to experiment with OpenTofu

GitLab provides [ways](https://docs.gitlab.com/ee/user/infrastructure/iac/terraform_state.html) to manage Terraform state, but I find the instructions riddled with uncertainty and fears that my shared state and plans might include passwords, and that they could be stolen from there.

I don't want to spend my brain cells on fears, so I want a **stateless OpenTofu** mode. I am fine with maintaining all YouTube ids in config files manually, and store all API passwords as CI/CD secrets. Is that possible?

**Comment by ghost** [comment]
Generally, no. State is at the heart of OpenTofu and it can't be easily disabled. Also, OpenTofu doesn't have a method of automatic reconciliation based on resource IDs unless the provider explicitly implements it. For example, if you have a resource and rename it, that will typically result in the resource being deleted and recreated, which is quite catastrophic for a YouTube video. OpenTofu works best if you manage everything with it, it doesn't do as well for partially-managed scenarios.

For your specific use case, Ansible may be a better tool if a YouTube integration exists for it. However, given my personal experience with the YouTube API and Google APIs in general, I'd be very careful because rate limits can cause problems if the integration doesn't handle the error codes properly. Updating descriptions of even hundreds of videos is likely less work than trying to build a stable, working integration.

**Reply by abitrolly** [reply]
As far as I know YouTube video ID can not be renamed, and can not be recreated, so the recreating resource scenario doesn't apply here. What a tool should do in the case of missing ID here, is to fail.

In the end I am going to write simple Python script for the integration, and probably ask Gemini to do it for me, but I was interested to have some engine that can detect changes needed and ask to apply them. Ansible is an option too, but it is a bit ugly on the UX, and I don't think it has interactive mode.

**Comment by apparentlymart** [comment]
Given the problem statement here, I'm assuming this involves a custom OpenTofu provider that integrates with an API provided by YouTube, which allows managing the metadata associated with a video. For the sake of this discussion, let's assume that provider is called "youtube" and that it has a resource type called `youtube_video_metadata`.

The problem statement also includes an allowance for managing the relationship between video id and associated metadata in a config file manually. Let's assume that's the config file in question is a `.tfvars` file with content like the following:

```hcl
videos = {
  "E4WlUXrJgy4" = {
    title       = "Cute cat"
    description = "My cute cat reclines on the sofa."
  }
}
```

This could then be used with an OpenTofu configuration like this, assuming the hypothetical provider previously discussed:

```hcl
variable "videos" {
  type = map(object({
    title       = string
    description = string
  }))
}

resource "youtube_video_metadata" "all" {
  for_each = var.videos

  title       = each.value.title
  description = each.value.description
}

import {
  for_each = var.videos

  id = each.key
  to = youtube_video_metadata.all[each.key]
}
```

Planning this configuration with a totally absent prior state would then first attempt to "import" all of the specified videos as instances of `youtube_video_metadata.all`, and if successful would then compare their imported state with the desired state described in the configuration and propose changes where any differ.

If you choose to apply that plan, you'd have a state file on local disk which you could immediately discard, because nothing in it is actually needed for ongoing work: if you plan again without that state file, OpenTofu would again "import" the videos before proposing changes.

Although I used the native syntax of `.tfvars` here, if starting with YAML were a hard requirement then you could first transcode from YAML to `.tfvars.json` -- which is really just a JSON file, without any special requirements other than the data structure matching the input variable type constraints -- and then pass _that_ as the input variables file.

It's not a _pretty_ solution, but I expect it would work for the contrived situation described here.
